File: packages/native/src/core/graph.zig
const std = @import("std");
const mem = std.mem;
const Allocator = mem.Allocator;
const ArrayListUnmanaged = std.ArrayListUnmanaged;
const StringHashMap = std.StringHashMap;

/// Dependency graph node
pub const GraphNode = struct {
    name: []const u8,
    dependencies: [][]const u8,
    dependents: [][]const u8,
};

/// Topological sort result
pub const TopoResult = struct {
    order: [][]const u8,
    has_cycle: bool,
    cycle_nodes: [][]const u8,

    pub fn deinit(self: *TopoResult, allocator: Allocator) void {
        for (self.order) |n| allocator.free(n);
        allocator.free(self.order);
        for (self.cycle_nodes) |n| allocator.free(n);
        allocator.free(self.cycle_nodes);
    }
};

/// Dependency graph builder
pub const DependencyGraph = struct {
    allocator: Allocator,
    nodes: StringHashMap(GraphNode),

    const Self = @This();

    pub fn init(allocator: Allocator) Self {
        return Self{
            .allocator = allocator,
            .nodes = StringHashMap(GraphNode).init(allocator),
        };
    }

    pub fn deinit(self: *Self) void {
        var iter = self.nodes.iterator();
        while (iter.next()) |entry| {
            self.allocator.free(entry.key_ptr.*);
            for (entry.value_ptr.dependencies) |dep| {
                self.allocator.free(dep);
            }
            self.allocator.free(entry.value_ptr.dependencies);
            for (entry.value_ptr.dependents) |dep| {
                self.allocator.free(dep);
            }
            self.allocator.free(entry.value_ptr.dependents);
            self.allocator.free(entry.value_ptr.name);
        }
        self.nodes.deinit();
    }

    /// Add node to graph
    pub fn addNode(self: *Self, name: []const u8, dependencies: [][]const u8) !void {
        const name_copy = try self.allocator.dupe(u8, name);

        var deps_copy = try self.allocator.alloc([]const u8, dependencies.len);
        for (dependencies, 0..) |dep, i| {
            deps_copy[i] = try self.allocator.dupe(u8, dep);
        }

        const node = GraphNode{
            .name = name_copy,
            .dependencies = deps_copy,
            .dependents = &[_][]const u8{},
        };

        try self.nodes.put(name_copy, node);
    }

    /// Build reverse edges (dependents)
    pub fn buildDependents(self: *Self) !void {
        var dependents_map = StringHashMap(ArrayListUnmanaged([]const u8)).init(self.allocator);
        defer {
            var iter = dependents_map.iterator();
            while (iter.next()) |entry| {
                entry.value_ptr.deinit(self.allocator);
            }
            dependents_map.deinit();
        }

        var iter = self.nodes.iterator();
        while (iter.next()) |entry| {
            const node_name = entry.key_ptr.*;

            for (entry.value_ptr.dependencies) |dep| {
                const result = try dependents_map.getOrPut(dep);
                if (!result.found_existing) {
                    result.value_ptr.* = ArrayListUnmanaged([]const u8){};
                }
                try result.value_ptr.append(self.allocator, node_name);
            }
        }

        iter = self.nodes.iterator();
        while (iter.next()) |entry| {
            if (dependents_map.get(entry.key_ptr.*)) |*list| {
                entry.value_ptr.dependents = try list.toOwnedSlice(self.allocator);
            }
        }
    }

    /// Topological sort with cycle detection (Kahn's algorithm)
    pub fn topologicalSort(self: *Self) !TopoResult {
        var in_degree = StringHashMap(usize).init(self.allocator);
        defer in_degree.deinit();

        var iter = self.nodes.iterator();
        while (iter.next()) |entry| {
            try in_degree.put(entry.key_ptr.*, entry.value_ptr.dependencies.len);
        }

        var queue = ArrayListUnmanaged([]const u8){};
        defer queue.deinit(self.allocator);

        iter = in_degree.iterator();
        while (iter.next()) |entry| {
            if (entry.value_ptr.* == 0) {
                try queue.append(self.allocator, entry.key_ptr.*);
            }
        }

        var result = ArrayListUnmanaged([]const u8){};

        while (queue.items.len > 0) {
            const node = queue.orderedRemove(0);
            try result.append(self.allocator, try self.allocator.dupe(u8, node));

            if (self.nodes.get(node)) |graph_node| {
                for (graph_node.dependents) |dependent| {
                    if (in_degree.getPtr(dependent)) |degree| {
                        degree.* -= 1;
                        if (degree.* == 0) {
                            try queue.append(self.allocator, dependent);
                        }
                    }
                }
            }
        }

        const has_cycle = result.items.len != self.nodes.count();
        var cycle_nodes = ArrayListUnmanaged([]const u8){};

        if (has_cycle) {
            iter = in_degree.iterator();
            while (iter.next()) |entry| {
                if (entry.value_ptr.* > 0) {
                    try cycle_nodes.append(self.allocator, try self.allocator.dupe(u8, entry.key_ptr.*));
                }
            }
        }

        return TopoResult{
            .order = try result.toOwnedSlice(self.allocator),
            .has_cycle = has_cycle,
            .cycle_nodes = try cycle_nodes.toOwnedSlice(self.allocator),
        };
    }

    /// Get all dependents of a node (transitive)
    pub fn getTransitiveDependents(self: *Self, node_name: []const u8) ![][]const u8 {
        var visited = StringHashMap(void).init(self.allocator);
        defer visited.deinit();

        var result = ArrayListUnmanaged([]const u8){};

        try self.dfsVisit(node_name, &visited, &result);

        return result.toOwnedSlice(self.allocator);
    }

    fn dfsVisit(
        self: *Self,
        node_name: []const u8,
        visited: *StringHashMap(void),
        result: *ArrayListUnmanaged([]const u8),
    ) !void {
        if (visited.contains(node_name)) return;

        try visited.put(node_name, {});

        if (self.nodes.get(node_name)) |node| {
            for (node.dependents) |dependent| {
                try result.append(self.allocator, try self.allocator.dupe(u8, dependent));
                try self.dfsVisit(dependent, visited, result);
            }
        }
    }

    /// Get all packages affected by changes (smart incremental builds)
    pub fn getAffectedPackages(self: *Self, changed_packages: []const []const u8) ![][]const u8 {
        var affected = StringHashMap(void).init(self.allocator);
        defer affected.deinit();

        var result = ArrayListUnmanaged([]const u8){};

        for (changed_packages) |changed| {
            if (!affected.contains(changed)) {
                try affected.put(changed, {});
                try result.append(self.allocator, try self.allocator.dupe(u8, changed));
            }

            const dependents = try self.getTransitiveDependents(changed);
            defer {
                for (dependents) |d| self.allocator.free(d);
                self.allocator.free(dependents);
            }

            for (dependents) |dependent| {
                if (!affected.contains(dependent)) {
                    try affected.put(dependent, {});
                    try result.append(self.allocator, try self.allocator.dupe(u8, dependent));
                }
            }
        }

        return result.toOwnedSlice(self.allocator);
    }

    /// Export graph as JSON for FFI
    pub fn toJson(self: *Self) ![]const u8 {
        var buffer = ArrayListUnmanaged(u8){};

        try buffer.appendSlice(self.allocator, "{\"nodes\":[");

        var iter = self.nodes.iterator();
        var first = true;
        while (iter.next()) |entry| {
            if (!first) try buffer.append(self.allocator, ',');
            first = false;

            try buffer.appendSlice(self.allocator, "{\"name\":\"");
            try buffer.appendSlice(self.allocator, entry.value_ptr.name);
            try buffer.appendSlice(self.allocator, "\",\"deps\":[");

            for (entry.value_ptr.dependencies, 0..) |dep, i| {
                if (i > 0) try buffer.append(self.allocator, ',');
                try buffer.append(self.allocator, '"');
                try buffer.appendSlice(self.allocator, dep);
                try buffer.append(self.allocator, '"');
            }

            try buffer.appendSlice(self.allocator, "],\"dependents\":[");

            for (entry.value_ptr.dependents, 0..) |dep, i| {
                if (i > 0) try buffer.append(self.allocator, ',');
                try buffer.append(self.allocator, '"');
                try buffer.appendSlice(self.allocator, dep);
                try buffer.append(self.allocator, '"');
            }

            try buffer.appendSlice(self.allocator, "]}");
        }

        try buffer.appendSlice(self.allocator, "]}");
        try buffer.append(self.allocator, 0); // null terminator

        return buffer.toOwnedSlice(self.allocator);
    }
};

File: crates/neex-core/src/graph.rs
//! Workspace Dependency Graph - Monorepo Support
//!
//! Features:
//! - Workspace discovery from package.json globs
//! - Directed dependency graph (petgraph)
//! - Topological sort for build order
//! - Cycle detection to prevent infinite builds
//! - Living graph updated by daemon

use anyhow::{anyhow, Result};
use petgraph::algo::{is_cyclic_directed, toposort};
use petgraph::graph::{DiGraph, NodeIndex};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};

/// A workspace package node
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkspaceNode {
    /// Package name from package.json
    pub name: String,
    /// Relative path from root
    pub path: PathBuf,
    /// Full path to package.json
    pub package_json_path: PathBuf,
    /// Package version
    pub version: Option<String>,
    /// Scripts available
    pub scripts: Vec<String>,
}

/// Root package.json structure
#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct RootPackageJson {
    name: Option<String>,
    workspaces: Option<WorkspacesConfig>,
}

/// Workspaces can be array or object with packages key
#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum WorkspacesConfig {
    Array(Vec<String>),
    Object { packages: Vec<String> },
}

/// Package-level package.json structure
#[derive(Debug, Deserialize)]
struct PackageJson {
    name: Option<String>,
    version: Option<String>,
    scripts: Option<HashMap<String, String>>,
    dependencies: Option<HashMap<String, String>>,
    #[serde(rename = "devDependencies")]
    dev_dependencies: Option<HashMap<String, String>>,
    #[serde(rename = "peerDependencies")]
    peer_dependencies: Option<HashMap<String, String>>,
}

/// Dependency Graph for workspace packages
pub struct DepGraph {
    /// The directed graph
    pub graph: DiGraph<WorkspaceNode, ()>,
    /// Map from package name to node index
    name_to_idx: HashMap<String, NodeIndex>,
    /// Root directory
    root: PathBuf,
}

impl DepGraph {
    /// Create a new empty dependency graph
    pub fn new() -> Self {
        Self {
            graph: DiGraph::new(),
            name_to_idx: HashMap::new(),
            root: PathBuf::new(),
        }
    }

    /// Build graph from a monorepo root directory
    pub fn from_root(root: impl AsRef<Path>) -> Result<Self> {
        let root = root.as_ref().to_path_buf();
        let mut graph = Self {
            graph: DiGraph::new(),
            name_to_idx: HashMap::new(),
            root: root.clone(),
        };

        // Step 1: Discover workspaces
        let workspaces = graph.discover_workspaces()?;

        // Step 2: Add nodes for each workspace
        for ws_path in &workspaces {
            if let Err(e) = graph.add_workspace_node(ws_path) {
                tracing::warn!("Failed to add workspace {:?}: {}", ws_path, e);
            }
        }

        // Step 3: Build dependency edges
        graph.build_edges()?;

        Ok(graph)
    }

    /// Discover workspace directories from root package.json
    fn discover_workspaces(&self) -> Result<Vec<PathBuf>> {
        let root_pkg_path = self.root.join("package.json");
        if !root_pkg_path.exists() {
            return Err(anyhow!("No package.json found in {:?}", self.root));
        }

        let content = std::fs::read_to_string(&root_pkg_path)?;
        let root_pkg: RootPackageJson = serde_json::from_str(&content)?;

        let patterns = match root_pkg.workspaces {
            Some(WorkspacesConfig::Array(arr)) => arr,
            Some(WorkspacesConfig::Object { packages }) => packages,
            None => return Ok(vec![]), // Not a monorepo
        };

        let mut workspaces = Vec::new();

        for pattern in patterns {
            // Handle glob patterns like "packages/*"
            let glob_pattern = self.root.join(&pattern).to_string_lossy().to_string();

            for entry in glob::glob(&glob_pattern)? {
                if let Ok(path) = entry {
                    if path.is_dir() && path.join("package.json").exists() {
                        workspaces.push(path);
                    }
                }
            }
        }

        Ok(workspaces)
    }

    /// Add a workspace node to the graph
    fn add_workspace_node(&mut self, ws_path: &Path) -> Result<NodeIndex> {
        let pkg_json_path = ws_path.join("package.json");
        let content = std::fs::read_to_string(&pkg_json_path)?;
        let pkg: PackageJson = serde_json::from_str(&content)?;

        let name = pkg.name.ok_or_else(|| anyhow!("Package has no name: {:?}", ws_path))?;
        let relative_path = ws_path.strip_prefix(&self.root).unwrap_or(ws_path).to_path_buf();

        let scripts: Vec<String> =
            pkg.scripts.map(|s| s.keys().cloned().collect()).unwrap_or_default();

        let node = WorkspaceNode {
            name: name.clone(),
            path: relative_path,
            package_json_path: pkg_json_path,
            version: pkg.version,
            scripts,
        };

        let idx = self.graph.add_node(node);
        self.name_to_idx.insert(name, idx);

        Ok(idx)
    }

    /// Build dependency edges between nodes
    fn build_edges(&mut self) -> Result<()> {
        // Collect all dependencies first (to avoid borrow issues)
        let mut edges: Vec<(String, String)> = Vec::new();

        for idx in self.graph.node_indices() {
            let node = &self.graph[idx];
            let pkg_content = std::fs::read_to_string(&node.package_json_path)?;
            let pkg: PackageJson = serde_json::from_str(&pkg_content)?;

            let source_name = node.name.clone();

            // Collect all dependency types
            let all_deps: Vec<String> = [
                pkg.dependencies,
                pkg.dev_dependencies,
                pkg.peer_dependencies,
            ]
            .into_iter()
            .flatten()
            .flat_map(|m| m.into_keys())
            .collect();

            for dep_name in all_deps {
                // Only add edge if dependency is in our workspace
                if self.name_to_idx.contains_key(&dep_name) {
                    edges.push((source_name.clone(), dep_name));
                }
            }
        }

        // Add edges (dependent -> dependency, so dependency comes first in topo sort)
        for (from_name, to_name) in edges {
            if let (Some(&from_idx), Some(&to_idx)) = (
                self.name_to_idx.get(&from_name),
                self.name_to_idx.get(&to_name),
            ) {
                // Edge direction: dependent -> dependency
                // This ensures dependencies come FIRST in topological order
                self.graph.add_edge(from_idx, to_idx, ());
            }
        }

        Ok(())
    }

    /// Check if graph has a cycle (circular dependency)
    pub fn has_cycle(&self) -> bool {
        is_cyclic_directed(&self.graph)
    }

    /// Get build order (topological sort)
    /// Returns packages in order they should be built (dependencies first)
    pub fn get_build_order(&self) -> Result<Vec<&WorkspaceNode>> {
        if self.has_cycle() {
            return Err(anyhow!("Circular dependency detected!"));
        }

        // toposort returns nodes in dependency order (dependencies last)
        // We need to reverse because we want dependencies first
        let sorted = toposort(&self.graph, None)
            .map_err(|_| anyhow!("Cycle detected during topological sort"))?;

        // Reverse: dependencies should come first
        let build_order: Vec<&WorkspaceNode> =
            sorted.into_iter().rev().map(|idx| &self.graph[idx]).collect();

        Ok(build_order)
    }

    /// Get packages affected by a change in the given package
    pub fn get_affected(&self, package_name: &str) -> Vec<&WorkspaceNode> {
        let Some(&start_idx) = self.name_to_idx.get(package_name) else {
            return vec![];
        };

        // BFS to find all packages that depend on this one
        let mut affected = vec![&self.graph[start_idx]];
        let mut visited = std::collections::HashSet::new();
        let mut queue = vec![start_idx];

        visited.insert(start_idx);

        while let Some(idx) = queue.pop() {
            // Find all packages that have an edge TO this node (dependents)
            for neighbor in self.graph.neighbors_directed(idx, petgraph::Direction::Incoming) {
                if visited.insert(neighbor) {
                    affected.push(&self.graph[neighbor]);
                    queue.push(neighbor);
                }
            }
        }

        affected
    }

    /// Get package by name
    pub fn get_package(&self, name: &str) -> Option<&WorkspaceNode> {
        self.name_to_idx.get(name).map(|&idx| &self.graph[idx])
    }

    /// Get all packages
    pub fn packages(&self) -> Vec<&WorkspaceNode> {
        self.graph.node_indices().map(|idx| &self.graph[idx]).collect()
    }

    /// Get package count
    pub fn package_count(&self) -> usize {
        self.graph.node_count()
    }

    /// Get edge count
    pub fn edge_count(&self) -> usize {
        self.graph.edge_count()
    }
}

impl Default for DepGraph {
    fn default() -> Self {
        Self::new()
    }
}

// Re-export old DependencyGraph for compatibility
pub type DependencyGraph = DepGraph;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cycle_detection() {
        let mut dep_graph = DepGraph::new();

        let idx_a = dep_graph.graph.add_node(WorkspaceNode {
            name: "A".into(),
            path: PathBuf::new(),
            package_json_path: PathBuf::new(),
            version: None,
            scripts: vec![],
        });
        let idx_b = dep_graph.graph.add_node(WorkspaceNode {
            name: "B".into(),
            path: PathBuf::new(),
            package_json_path: PathBuf::new(),
            version: None,
            scripts: vec![],
        });

        // A -> B
        dep_graph.graph.add_edge(idx_a, idx_b, ());
        assert!(!dep_graph.has_cycle(), "Should not detect cycle yet");

        // B -> A (creates cycle!)
        dep_graph.graph.add_edge(idx_b, idx_a, ());
        assert!(dep_graph.has_cycle(), "Should detect cycle!");
    }

    #[test]
    fn test_topological_sort() {
        let mut dep_graph = DepGraph::new();

        // utils <- ui <- web
        // utils has no deps, ui depends on utils, web depends on ui and utils

        let utils_idx = dep_graph.graph.add_node(WorkspaceNode {
            name: "@my/utils".into(),
            path: PathBuf::from("packages/utils"),
            package_json_path: PathBuf::new(),
            version: None,
            scripts: vec![],
        });
        dep_graph.name_to_idx.insert("@my/utils".into(), utils_idx);

        let ui_idx = dep_graph.graph.add_node(WorkspaceNode {
            name: "@my/ui".into(),
            path: PathBuf::from("packages/ui"),
            package_json_path: PathBuf::new(),
            version: None,
            scripts: vec![],
        });
        dep_graph.name_to_idx.insert("@my/ui".into(), ui_idx);

        let web_idx = dep_graph.graph.add_node(WorkspaceNode {
            name: "@my/web".into(),
            path: PathBuf::from("packages/web"),
            package_json_path: PathBuf::new(),
            version: None,
            scripts: vec![],
        });
        dep_graph.name_to_idx.insert("@my/web".into(), web_idx);

        // ui -> utils (ui depends on utils)
        dep_graph.graph.add_edge(ui_idx, utils_idx, ());
        // web -> ui (web depends on ui)
        dep_graph.graph.add_edge(web_idx, ui_idx, ());
        // web -> utils (web depends on utils)
        dep_graph.graph.add_edge(web_idx, utils_idx, ());

        let order = dep_graph.get_build_order().unwrap();
        let names: Vec<&str> = order.iter().map(|n| n.name.as_str()).collect();

        // utils must come before ui, ui must come before web
        let utils_pos = names.iter().position(|&n| n == "@my/utils").unwrap();
        let ui_pos = names.iter().position(|&n| n == "@my/ui").unwrap();
        let web_pos = names.iter().position(|&n| n == "@my/web").unwrap();

        assert!(utils_pos < ui_pos, "utils should come before ui");
        assert!(ui_pos < web_pos, "ui should come before web");
    }

    #[test]
    fn test_affected_packages() {
        let mut dep_graph = DepGraph::new();

        let utils_idx = dep_graph.graph.add_node(WorkspaceNode {
            name: "@my/utils".into(),
            path: PathBuf::from("packages/utils"),
            package_json_path: PathBuf::new(),
            version: None,
            scripts: vec![],
        });
        dep_graph.name_to_idx.insert("@my/utils".into(), utils_idx);

        let ui_idx = dep_graph.graph.add_node(WorkspaceNode {
            name: "@my/ui".into(),
            path: PathBuf::from("packages/ui"),
            package_json_path: PathBuf::new(),
            version: None,
            scripts: vec![],
        });
        dep_graph.name_to_idx.insert("@my/ui".into(), ui_idx);

        let web_idx = dep_graph.graph.add_node(WorkspaceNode {
            name: "@my/web".into(),
            path: PathBuf::from("packages/web"),
            package_json_path: PathBuf::new(),
            version: None,
            scripts: vec![],
        });
        dep_graph.name_to_idx.insert("@my/web".into(), web_idx);

        // ui -> utils
        dep_graph.graph.add_edge(ui_idx, utils_idx, ());
        // web -> ui
        dep_graph.graph.add_edge(web_idx, ui_idx, ());

        // If utils changes, both ui and web are affected
        let affected = dep_graph.get_affected("@my/utils");
        let names: Vec<&str> = affected.iter().map(|n| n.name.as_str()).collect();

        assert!(names.contains(&"@my/utils"));
        assert!(names.contains(&"@my/ui"));
        assert!(names.contains(&"@my/web"));
    }
}
